### DISCUSSION TITLE
[HowTo] External Ceph cluster provided by Proxmox and managed by Rook with RGW object storage

### DISCUSSION THREAD

**Discussion by alex-ioma** [rook/rook#14728] [Q&A]
Hi all, 

This is my current setup: 
- Proxmox VE cluster (3 nodes) with Ceph cluster provided by PVE install
- ceph_cluster and ceph_public physically separated networks (dedicated hardware interfaces for PVE nodes)
- k8s based on (rke2) with 3 master + 3 workers node (VMs in PVE)
- k8s workers node has a eth interface that can reach the ceph_public network
- current mgr configuration

```bash
:~# ceph osd pool ls
.mgr
ceph-pool
cephfs_data
cephfs_metadata

:~# ceph osd tree
ID  CLASS  WEIGHT   TYPE NAME       STATUS  REWEIGHT  PRI-AFF
-1         5.23979  root default                             
-3         1.74660      host pve01                           
 0   nvme  1.74660          osd.0       up   1.00000  1.00000
-5         1.74660      host pve02                           
 1   nvme  1.74660          osd.1       up   1.00000  1.00000
-7         1.74660      host pve03                           
 2   nvme  1.74660          osd.2       up   1.00000  1.00000

:~# ceph mgr module ls
MODULE                           
balancer           on (always on)
crash              on (always on)
devicehealth       on (always on)
orchestrator       on (always on)
pg_autoscaler      on (always on)
progress           on (always on)
rbd_support        on (always on)
status             on (always on)
telemetry          on (always on)
volumes            on (always on)
iostat             on            
nfs                on            
restful            on            
alerts             -             
influx             -             
insights           -             
localpool          -             
mirroring          -             
osd_perf_query     -             
osd_support        -             
prometheus         -             
selftest           -             
snap_schedule      -             
stats              -             
telegraf           -             
test_orchestrator  -             
zabbix             -             
```

The goal: 
- Deploy rook to manage the external cluster providing RWX StorageClass backed by a new dedicated cephfs pool (say `rook-fs-data`) and object storage with other dedicated pools (say rook-obj.*)
- Rook should deploy RGW on the k8s cluster and expose it through an ingress (nginx / traefik + LE cert via cert-manager)
- Rook should ideally provide the dashboard that connects to the mgr and expose it via ingress
- Rook should act as a relay for prometheus (already installed) monitoring

Current status: 
- I've seen references to the `create-cluster-resources.py` and `import-external-cluster.sh` scripts to be used in conjunction with `common-external.yaml` and `cluster-external.yaml` but I haven't found any specific guide on how to use `cluster-external-managed.yaml`
- It is not clear to me if the above scenario is achievable or RGW will always need to be installed on the external cluster

Before pressing ahead and break things, I though about opening a dedicated issue hoping to find guidance, avoid silly mistakes and also document this type of application for future use (since PVE is not going to support object storage directly anytime soon). (Not breaking a few clusters to get this right would definitely be a plus.)

I really appreciate the support and thank in advance anyone who could point me to the right documentation and/or provide how-to suggestions.

**Comment by parth-gr** [comment]
> Deploy rook to manage the external cluster providing RWX StorageClass backed by a new dedicated cephfs pool (say rook-fs-data) and object storage with other dedicated pools (say rook-obj.*)

use --cephfs-filesystem-name and --rgw-endpoint flag while running the python script

> Rook should deploy RGW on the k8s cluster and expose it through an ingress (nginx / traefik + LE cert via cert-manager)
Use the doc https://rook.io/docs/rook/latest-release/CRDs/Cluster/external-cluster/external-cluster/#connect-to-an-external-object-store

External ceph cluster should deploy it, we provide object bucket through Rook,
@BlaineEXE iirc we dont create an extra service for rgw


>Rook should ideally provide the dashboard that connects to the mgr and expose it via ingress
Rook should act as a relay for prometheus (already installed) monitoring

Rook would be creating a secret and service that can be utilised to use the ceph dashboard, need to check we dont document it but would be similar to internal mode

>It is not clear to me if the above scenario is achievable or RGW will always need to be installed on the external cluster

We can have any type of storage no need to have all of them at the same time

>cluster-external-managed.yaml

Not required we will probably remove it from the repo

**Reply by alex-ioma** [reply]
Thank you @parth-gr!

To clarify: since PVE does not deploy RGW by default, I would need to add it manually to the cluster, say following [the ceph doc](https://docs.ceph.com/en/reef/install/manual-deployment/#manually-installing-radosgw) so that rook can later interface to it? 

Does the same apply for a dedicated pool to be later used as a RWX storageClass? Does that also need to be manually created on source Ceph prior to running the python script? 

I was under the impression that Rook could have provided individual components of Ceph, such as RGW or MDS only, even when mgr, mon, osd and pools are coming from a source cluster, as described [here](https://rook.io/docs/rook/latest-release/Getting-Started/example-configurations/#cluster-crd) using `cluster-external-management.yaml`.

**Reply by alex-ioma** [reply]
PS. My goal would have been to setup a HA RGW with three nodes taking advantage rook simple deployment k8s and avoiding the extra overheads of manual deployment since no RGW is yet deployed on the source cluster.

**Reply by alex-ioma** [reply]
@travisn could you please weigh in on this issue? I've noticed that last commit on `cluster-external-management.yaml ` is yours.

**Reply by alex-ioma** [reply]
@parth-gr To be more precise, my doubts arises from reading rook docs for [design/ceph-external-cluster.md](https://github.com/rook/rook/blob/master/design/ceph/ceph-external-cluster.md#object-storage-rgw) at master.

**Comment by BlaineEXE** [comment]
It's been a while since I have validated this, but at one point Rook was able to deploy RGWs in Kubernetes that reference a Ceph cluster outside of Kubernetes. Doing so requires giving Rook admin permissions to the Ceph cluster to allow it to create and configure the pools that back the CephObjectStore. To do so, you need a CephCluster in external mode and a CephObjectStore in internal mode, which is just a standard deploy (don't set externalRgwEndpoints).

I'm not sure if this fully answers all the questions you have, but I think it will get you moving at least a little.

**Reply by parth-gr** [reply]
@BlaineEXE lets dicuss on this, if the cephcluster mode is external, does the reconciliation of object store won't still obey the external mode path?

**Reply by BlaineEXE** [reply]
It sounds like you are understanding that the CephCluster external mode and CephObjectStore external mode would be different but are confused by how that would work in practice. 

It's helpful to think of any Ceph cluster (internal, external, Rook, Proxmox, cephadm, or any other configuration) as a building block that is the hub for Ceph-related storage. As long as I have access to that cluster and have a Ceph admin key, I can create anything I need and connect anything I want to that Ceph hub.

In the Internal-Internal (CephCluster-CephObjectStore) case, Rook creates the hub, already has the admin key, and can therefore bootstrap the RGWs.

The Internal-External case doesn't really have a purpose I can think of, but it could be done just for giggles by setting up an internal CephObjectStore and then creating a second external CephObjectStore that references the first.

The External-External case is the one we usually focus on. The benefits of this are that we have a clear separation between Ceph cluster management (via cephadm/Proxmox) and Rook which is set up to use a consumption-only model. It means that Rook only needs minimal read permissions to the Ceph cluster, allowing multi-tenancy and strong isolation between consumer/Rook peers.

The External-Internal case is what Alex is asking about. At first glance, it might seem natural to think that this wouldn't work, but that's where the idea of Ceph as a hub comes in. Rook can be given an admin key, which allows it to provision the CephObjectStore via the internal path. The admin key allows Rook permissions to create/modify pools, create cephx keys, and whatever else it needs. It also has permissions to use radosgw-admin to bootstrap the RGW configs. We don't test this configuration in CI, and I haven't tested it in minikube in over 2 years (maybe more than 3). It's possible that there are 1 or more code checks in Rook that might stop this from working, but those would be somwhat artificial limitations. The theory of this use-case is still viable.

**Reply by alex-ioma** [reply]
Thank you @BlaineEXE for the clarification. Indeed this is the setup I was looking into and, truth to be told, after spending the last few days tinkering with ceph (or the proxmox's version thereof) I understand why it might be be advisable to run the external-internal scenario in production or why rook might want to put in place code-checks and limitation for such a use case. 

More details in thread below.

**Comment by alex-ioma** [comment]
Thank you @BlaineEXE. I'm still not clear to me which would be the best path forward now. 

Ideally, I would not want to to pass full admin to Rook since the external Ceph is also used as the the backend storage for Proxmox VMs on dedicated pools. Here my concerns is not much security but rather a maintainer erroneously configuring something on Rook and affecting the underlying infrastructure that holds Kubernetes/Rook. 

That said, my case can be extended beyond Proxmox itself and my main issues are: 
- I have an external ceph that has been deployed without RGW
- I'd like to provide RGW components through Rook to take advantage of easier HA configurations. (Would it be possible to easily spin up 3 node RGWs otherwise and make sure there is no tls issue with Rook? How about exposing s3 outside the cluster? This is why I was in favor of RGW being in Rook). 
- I'd also like to easily expose the dashboard through ingress (since my underlying nodes are not exposed and I would not want to manage HAProxy or similar).
- I'd like to gather metrics through Proxmox (which is running in k8s) - this seems the easiest of the task. 

What I'd like to try is to run python script with this `config.ini` after pre-creating all necessary pools on external ceph. 

```
[Configurations]
format = bash
namespace = rook-ceph
cephfs-filesystem-name = rook-fs
rbd-data-pool-name = rook
rgw-pool-prefix = rook
```

Then spin up a `cluster-external.yaml` and try as you suggested with an internal CephObjectStore. But would RGW keyring have correct privileges/caps at this point?

Yet I wonder, is this the correct/best way to achieve my goals? Or would is there a simpler solution to the task and perhaps starting 3 RGW on the external ceph is just as easy and better documented than the approach above? 

I understand this might be a chicken-and-egg situation and I'd really value your opinion on the matter. 

Thank you in advance.

**Reply by BlaineEXE** [reply]
I don't think it would be appropriate for Rook maintainers to have a strong statement about what is best. Every user is different and has different requirements and abilities. 

You have pretty concisely identified what to me seem like the major pros/cons between Rook running RGWs for your cluster and spinning up RGWs for your cluster yourself. 

As a middle ground approach, it might be possible to pre-provision the pools that Rook's CephObjectStore would need, allowing you to give Rook fewer permissions. However, RGW doesn't have a permanently fixed set of pools or configs. Ceph RGW software evolves over time, and there is risk that you may have to resolve integration issues when there are major version updates for Rook and/or Ceph on occasion. You would largely have to chart that path for yourself since it's not tested or supported -- only theoretically possible.

**Reply by alex-ioma** [reply]
Thank you. I'm not much concern about evolvability since I would bucket changes are typically link to major version upgrade and this will be managed with care to check for breaking changes. 

I understand the importance on not sharing a position on the "best" approach, since this ultimately depends also on personal preference and the tools each maintainer is more accustom to. However, this was more a request for suggestion on methods that could better suite all people trying to accomplish this task that may not have extensive knowledge on ceph (and therefore rook) since documentation on this is there but it seems there is not a clear path forward that can be easily followed. Even a personal preference would be appreciated and respected as such.

**Reply by BlaineEXE** [reply]
I guess the challenge is that PVE doesn't support RGWs, as you've noted.

You can look into the technical stuff from my comment here, explaining some differences/benefits between different scenarios here: https://github.com/rook/rook/discussions/14728#discussioncomment-10694658

If I were setting this up for myself, I would probably lean towards running RGWs with the Ceph cluster itself.

The core reasons for me to use this model are:

- The External-Internal model can suffer from unexpected challenges with Rook's usage of `radosgw-admin` to bootstrap/modify RGW configs. The Ceph team that develops `radosgw-admin` doesn't ensure forward compatibility, and that would almost certainly cause problems from time to time. Using radosgw-admin manually via PVE seems easier than hitting a bug in Rook and needing to fix Golang code than just running some commands myself.
- The External-Internal model isn't tested in CI, so I don't have strong guarantees things will work on an ongoing basis.
- While I trust that Rook is safe, it's still an even safer architecture to keep the admin key only with the Ceph cluster management.
- Keeping management of the Ceph cluster and Rook cluster separate sets clear boundaries around what is managed where, keeping the mental overhead for understanding each system lower, and easier
- Managing HAProxy is a burden, but it's a well-known system and likely a smaller amount of maintenance long-term compared to the issues in my first bullet point. I've found HAProxy to be reasonably easy to set up in my limited experiences.

From a brief look at the PVE docs, it seems that setting up RGW is largely an exercise left up to the reader. If that's the case, there's also nothing preventing you from creating your own RGW deployments/statefulsets to run in K8s, instead of in PVE's (virtual?) environment.

**Reply by alex-ioma** [reply]
Hi @BlaineEXE, thanks again for the very detailed answers and advices. Really appreciated it. 

I agree with every point you mentioned, and from my limited experience with Ceph, the external-internal scenario is definitely not suited for production.

Also, Proxmox's implementation of Ceph is indeed lacking in many areas, which leaves much of the troubleshooting to the user, especially when it comes to object storage. I don't blame their team for not supporting RGW capabilities, but their customized implementation of Ceph makes it difficult to achieve a solid setup. Even before your previous post, I had decided to deploy RGW at the Proxmox level and have encountered several challenges in getting a working configuration. Half of the commands documented in the official Ceph Reef documentation simply don’t work in Proxmox’s version of RGW. Additionally, the fact that Proxmox does not use an orchestrator makes it even harder to deploy additional components. Even the dashboard throws unlogged 500 errors with RGW and causes the MGR to crash if the wrong tab is opened in the browser.

Thus, my original idea was to go for the external-internal strategy and let Rook, along with the well-known (at least to me) Kubernetes components, handle all the necessary dependencies for RGW. However, this approach was a no-go for the reasons you outlined.

Setting up HAProxy (especially in Kubernetes) is not a big deal after all, and I'll likely follow your suggestion and go this route if I can get the underlying RGW Ceph cluster to a stable state.

To your last point, the reasons I was trying to avoid setting up a separate cluster for RGW are:

- My current setup includes a 3-node Proxmox + Ceph cluster with 2 NVMe drives per node. One was intended for Proxmox HA configuration with Ceph, while the other was to be passed to the VM running Kubernetes, exclusively for MongoDB replicas.
- Object storage should be provided to the cluster.
- Given this setup, the goal was to optimize the use of the disks (even though they are NVMe drives) to maximize lifespan.
- I wanted to maintain the HA function of Proxmox for other VMs that might be running alongside the Kubernetes cluster.

The logical idea was then to use the pre-existing Ceph cluster as backing for Rook. While Rook can be deployed in minutes via Helm, I had hoped that this external scenario (even in fully external) wouldn't take so much time.

I'll keep you posted if I succeed and will share a guide for future wanderers like me. However, I’m seriously considering getting rid of Proxmox and running bare metal with all of Ceph directly managed by Rook (FYI).

**Reply by alex-ioma** [reply]
Hi @BlaineEXE, I've followed the implementation guide after setting up my cluster. Everything seems fine on k8s and rook, but object storage doesn't work - after deploying cluster, object-store and sc and`deploy/examples/object-bucket-claim-delete.yaml` the bucket does not get created. 

I realize this are not a log of information but any suggestion on where to look to troubleshoot would be appreciated. 

PS. I don't see any aooarent issue in RGW logs. 

Thanks.

**Reply by alex-ioma** [reply]
I've corrected some of the configs but currently I have issues with the export import. Is it correct that having 3 mons all with v2-port enabled the config that is generated is similar to this:
```
[{"clusterID":"rook-ceph","monitors":["10.2.252.12:6789","10.2.252.13:6789","10.2.252.11:3300"],"cephFS":{"netNamespaceFilePath":"","subvolumeGroup":"","kernelMountOptions":"","fuseMountOptions":""},"rbd":{"netNamespaceFilePath":"","radosNamespace":""},"nfs":{"netNamespaceFilePath":""},"readAffinity":{"enabled":false,"crushLocationLabels":["kubernetes.io/hostname","topology.kubernetes.io/region","topology.kubernetes.io/zone","topology.rook.io/chassis","topology.rook.io/rack","topology.rook.io/row","topology.rook.io/pdu","topology.rook.io/pod","topology.rook.io/room","topology.rook.io/datacenter"]},"namespace":"rook-ceph"}]
```

should the exported value be different and list all mons?
```
export ROOK_EXTERNAL_CEPH_MON_DATA=pve01=10.2.252.11:3300
```

I feel I'm close to a solution but cannot get things to work.

Best.

**Reply by BlaineEXE** [reply]
@parth-gr might be the best person to help with the recent comment. I think 6789 is the v1 port, so `["10.2.252.12:6789","10.2.252.13:6789","10.2.252.11:3300"]` looks concerning to me, but I may not have all the info.

**Reply by parth-gr** [reply]
If any one endpoint has v2, rook should reconcile to update all the endpoints, https://github.com/rook/rook/blob/master/pkg/operator/ceph/cluster/cluster_external.go#L119

Can you try operator restart?

**Reply by alex-ioma** [reply]
I already did before posting. 
But I did have some issue in my configuration. 

However there is something going very wrong in the process (and it might be my fault that I can see yet). 

In short, I'm able to export-import and kubectl apply the external cluster manifest. 

I then see the cluster object as correctly deployed and with health_ok. However I'm not able to deploy anything further. 

More specifically, object-external.yaml / sc / object-bucket-delete are not able to create object on ceph. 

After further inspecting the rook cluster, it seems that the external cluster is not really reacheable or controllable. 

Running a `ceph -s` from the toolbox reports: cluster not found. 

I'd like to provide better details but I'm a bit lost on how to troubleshoot. Can you please shad some light on how to move forward. 

I also tried to fully wipe rook and the related CRDs k8s and recreate everything from scratch - somehow the new clusters see the older object bucket claim and complaints that the resources was still in deletion - which is strange. 

As a side notes, I'm deploy the operator via helm and the cluster via manifest. But that shouldn't make a difference. 

Thanks in advance.

**Reply by parth-gr** [reply]
Can you share the specific logs what is not getting deleted or re-used,
If you have added some data to your application then you have to remove it manually

**Reply by alex-ioma** [reply]
Thanks @parth-gr . 

Of course the misconfigurations was on my end. I had issues both in the external cluster where I messed up with zones and realms (while I didn't want to embark on a multi-tenant cluster, I had erroneously set the realm) and with the `provider-export`. Brutally fast copy-past-ing I left behind the first config "export" from the script that was being printed in a different format and, as such, left behind. 

To clarify for future user, this is part of the values that needs to be copied with the rest of the python script output: 

```
export ARGS="[Configurations]
k8s-cluster-name = xxx
namespace = rook-ceph
rgw-pool-prefix = default
v2-port-enable = True
format = bash
cephfs-filesystem-name = cephfs
cephfs-metadata-pool-name = cephfs_metadata
cephfs-data-pool-name = cephfs_data
rbd-data-pool-name = rook
rgw-endpoint = 10.2.252.11:7480
monitoring-endpoint = 10.2.252.11,10.2.252.12,10.2.252.13
monitoring-endpoint-port = 9283
"
export NAMESPACE=rook-ceph
```
Beware of the first `ARGS`.

Since, we are at it though. After I added the corrected/full configuration with the importa script: 
1. I see mon endpoints being correctly reconfigure on port 3300 within `rook-ceph-csi-config` configMap, but I still see the issue within this secret `rook-ceph-mon-endpoints`
2. Can you please confirm that now seeing mon/mgr info on this command is by design?

```bash
❯ k get -n rook-ceph cephclusters.ceph.rook.io
NAME                 DATADIRHOSTPATH   MONCOUNT   AGE   PHASE       MESSAGE                          HEALTH      EXTERNAL   FSID
rook-ceph-external                                21h   Connected   Cluster connected successfully   HEALTH_OK   true       dfa7e83c-0c17-4428-95eb-ba5d4e7a0dfd
❯ k describe -n rook-ceph cephclusters.ceph.rook.io rook-ceph-external
Name:         rook-ceph-external
Namespace:    rook-ceph
Labels:       <none>
Annotations:  <none>
API Version:  ceph.rook.io/v1
Kind:         CephCluster
Metadata:
  Creation Timestamp:  2024-09-28T01:22:35Z
  Finalizers:
    cephcluster.ceph.rook.io
  Generation:        2
  Resource Version:  56526379
  UID:               6e384daa-eb07-4a6d-87df-a2aa52ded091
Spec:
  Ceph Version:
  Cleanup Policy:
    Sanitize Disks:
  Crash Collector:
    Disable:  true
  Csi:
    Cephfs:
    Read Affinity:
      Enabled:  false
  Dashboard:
  Disruption Management:
  External:
    Enable:  true
  Health Check:
    Daemon Health:
      Mon:
        Interval:  45s
      Osd:
      Status:
  Log Collector:
  Mgr:
  Mon:
  Monitoring:
  Network:
    Multi Cluster Service:
  Security:
    Key Rotation:
      Enabled:  false
    Kms:
  Storage:
    Flapping Restart Interval Hours:  0
    Store:
Status:
  Ceph:
    Capacity:
      Bytes Available:  5755670884352
      Bytes Total:      5761136590848
      Bytes Used:       5465706496
      Last Updated:     2024-09-28T23:10:51Z
    Fsid:               dfa7e83c-0c17-4428-95eb-ba5d4e7a0dfd
    Health:             HEALTH_OK
    Last Checked:       2024-09-28T23:10:51Z
    Versions:
      Mds:
        ceph version 18.2.2 (e9fe820e7fffd1b7cde143a9f77653b73fcec748) reef (stable):  3
      Mgr:
        ceph version 18.2.2 (e9fe820e7fffd1b7cde143a9f77653b73fcec748) reef (stable):  3
      Mon:
        ceph version 18.2.2 (e9fe820e7fffd1b7cde143a9f77653b73fcec748) reef (stable):  3
      Osd:
        ceph version 18.2.2 (e9fe820e7fffd1b7cde143a9f77653b73fcec748) reef (stable):  3
      Overall:
        ceph version 18.2.2 (e9fe820e7fffd1b7cde143a9f77653b73fcec748) reef (stable):  15
      Rgw:
        ceph version 18.2.2 (e9fe820e7fffd1b7cde143a9f77653b73fcec748) reef (stable):  3
  Conditions:
    Last Heartbeat Time:   2024-09-28T17:24:14Z
    Last Transition Time:  2024-09-28T17:24:14Z
    Message:               Attempting to connect to an external Ceph cluster
    Reason:                ClusterConnecting
    Status:                True
    Type:                  Connecting
    Last Heartbeat Time:   2024-09-28T23:10:51Z
    Last Transition Time:  2024-09-28T01:22:36Z
    Message:               Cluster connected successfully
    Reason:                ClusterConnected
    Status:                True
    Type:                  Connected
  Message:                 Cluster connected successfully
  Phase:                   Connected
  State:                   Connected
Events:                    <none>
```

**Reply by alex-ioma** [reply]
Also, I was getting an error that the `ObjectBucketClaim` CR had issues with the definition. 

```
W0928 23:27:09.273183       1 reflector.go:561] pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/reflector.go:243: failed to list *v1alpha1.ObjectBucketClaim: the server could not find the requested resource (get objectbucketclaims.objectbucket.io)
E0928 23:27:09.273216       1 reflector.go:158] "Unhandled Error" err="pkg/mod/k8s.io/client-go@v0.31.0/tools/cache/reflector.go:243: Failed to watch *v1alpha1.ObjectBucketClaim: failed to list *v1alpha1.ObjectBucketClaim: the server could not find the requested resource (get objectbucketclaims.objectbucket.io)" logger="UnhandledError"
2024-09-28 23:28:03.616394 I | ceph-bucket-notification: ObjectBucketClaim "default/ceph-delete-bucket" resource did not create the bucket yet. will retry
```
Could it be possible that helm deployment might miss some CRDs related to RGW?

Without destroying the cluster, it was sufficient to `kubectl apply` the CRDs git manifest and everything start working. 

So I finally have a successfully deployed bucket.

**Reply by alex-ioma** [reply]
UPDATED: 
- I've tried recreating everything in k8s from scratch via helm
- External cluster deployed successfully and Bucket Claim works as expected
- I'm still seeing the port error for the additional monitors which are referenced to port `6789` and not as `3300`?

Thanks.

**Comment by alex-ioma** [comment]
@parth-gr @BlaineEXE on a separate note, I was reading this: 

> `externalRgwEndpoints`: [...] Multiple endpoints can be given, but for stability of ObjectBucketClaims, we highly recommend that users give only a single external RGW endpoint that is a load balancer that sends requests to the multiple RGWs.

Since my k8s VM have a dedicated `eth` interface that taps into the ceph public network and can reach both mons and RGWs (which are uniquely exposed on that network - isolated from the rest of the LANs), I would like to avoid any LB on proxmox nodes. Thus, I was considering about deploying an internal HAProxy within k8s rook-ceph namespace so that a single IP can be passed as `object-external`.

Do you think this could work well? Do you see any performance / security issues? 

I was considering something like: 
```yaml
---
apiVersion: v1
kind: ConfigMap
metadata:
  name: rgw-internal-lb-config
  namespace: rook-ceph
data:
  haproxy.cfg: |
    frontend rgw-frontend
      bind *:80
      mode http
      default_backend rgw-backend
    backend rgw-backend
      mode http
      balance roundrobin
      option httpchk GET /
      http-check expect status 200
      cookie SERVERID insert indirect nocache
      server rgw1 10.2.252.11:7480 check cookie rgw1
      server rgw2 10.2.252.12:7480 check cookie rgw2
      server rgw3 10.2.252.13:7480 check cookie rgw3
```
and then a couple of pods in a very simple deployment and a ClusterIP service in front. 

Alternatively, an external LXC container could be easy enough so that the LB resides outside k8s.

Also, would sticky connection be counter-productive / useless with object storage APIs?

Thanks again for all the help.

**Reply by BlaineEXE** [reply]
I once prototyped using a service with no selectors and an endpointslice to point to external RGWs: https://kubernetes.io/docs/concepts/services-networking/service/#services-without-selectors

It works just fine, after trial and error to get everything configured properly (having consistent ports will help your use case), it works very similarly to a normal service. IIRC, there is no ability for k8s to detect/report endpoint readiness though, which is the main downside.

This simple approach would make a LB unnecessary though, which could be a win.

Otherwise, you can create a single externalname service for each endpoint and then a LB on top I think.

**Reply by alex-ioma** [reply]
Hi @BlaineEXE thank you for the response. 

You are indeed correct. EndpointSlices / externalEndpoints do not have health-checks in k8s (as they can only reflect the status of the heath checks of pod they point to). Hence the service won't be able to monitor the status of the endpoint. 

The EP Slices are a really neat solution but, unfortunately, won't be able to satisfy the requirements of acting as an effective HA LB - in case a RGW goes down, the service will still send traffic to it. 

So I ended up going with the internal HA Proxy solution, which works (although a bit of an overkill). I've simplified the config as follows - trying to minimize the logging abuse the health checks generates (Ceph RGW logs with an error each health check requests and I coun't find a better way to setup this up): 
```
    frontend rgw-frontend
        bind *:80
        mode http
        default_backend rgw-backend
    backend rgw-backend
        mode http
        balance roundrobin
        server rgw1 10.2.252.11:7480 check inter 1000 rise 1 fall 1
        server rgw2 10.2.252.12:7480 check inter 1000 rise 1 fall 1
        server rgw3 10.2.252.13:7480 check inter 1000 rise 1 fall 1
```

Fast forward to today. After many trial and error I finally make the whole solution works although I encountered many challenges. The main issues was the following: 
- Since rook's `object-external` works best with a single IP i've tired to point rook to the internal service by specifying hostname in the config and in replacing the RGW ip address in the import scrip. 
- This kept failing and did not allow me to connect to the RGW
- Turned out that the issue was in the `CephObjectStore` object, which appears to require a static ip. 

*Incorrect* Object Store:
```yaml
apiVersion: ceph.rook.io/v1
kind: CephObjectStore
metadata:
  name: external-store
  namespace: rook-ceph
spec:
  gateway:
    port: 7480
    externalRgwEndpoints:
      - hostname: rgw-internal-lb
```

To circumnavigate this, the solution was to: 
- Fix the `clusterIP` of the internal service 
- Specifying such ip (instead of the service name, as you would in a normal k8s scenario) in all rook's manifests referencing the RGW endpoint

Working solution: 
```yaml
apiVersion: v1
kind: Service
metadata:
  name: rgw-internal-lb
  namespace: rook-ceph
spec:
  type: ClusterIP
  clusterIP: 10.43.2.10
  # additional specs
---
apiVersion: ceph.rook.io/v1
kind: CephObjectStore
metadata:
  name: external-store
  namespace: rook-ceph
spec:
  gateway:
    port: 7480
    externalRgwEndpoints:
      - hostname: 10.43.2.10
```

In the next days I'll try to post a more complete description of the working solution but I wanted to start sharing this as there might be an additional check / lock from rook that prevents a simpler solution to be working. 

In other words, would it be possible to only specify the internal service name as RGW endpoint?

I understand that having an internal LB for RGW is a bit of unconventional solution, yet, in my scenario seemed to be a logic approach as I'm trying to keep the underlying Proxmox / Ceph cluster as clean as possible.

Thanks for the suppor. Comments are also welcome.

**Reply by appcoders** [reply]
Hi @alex-ioma,
I have been struggling for a week to get ceph-csi (rbd) and now rook running with my Proxmox ceph cluster. I'm using the external method and followed closely all the steps. The Connection is fine, but PVC creation is stuck and runs in timeout. In the logs, everything looks fine, but nothing happens after the log entry "setting image options". Do you have any hints? I am on ceph 17.2.7 and proxmox 8.3.2.  Thanks :-)

**Reply by appcoders** [reply]
And short addition logs are attached already here: https://github.com/ceph/ceph-csi/issues/5073

**Reply by Madhu-1** [reply]
@appcoders i replied on the ceph-csi thread, Let debug it on the ceph-csi issue.

**Reply by alex-ioma** [reply]
Hi @appcoders I am on Reef and all working as expected - besides dashboard (but this is more on the ceph side and not rook). I recall some differences with Quincy, but the setup should be working. 

I agree it should be better to debug on your issue.
